File: codewars/301-400/memoized-log-cutting.py
m = [ 0 for _ in range(3000) ]
op = []

def cut_log(p, n):
    global m
    global op
    if (p != op):
        m = [ 0 for _ in range(3000) ]
        op = p
    if (n == 0):
        return 0
    if (m[n] != 0):
        return m[n]
    q = -1
    for i in range(1, n + 1):
        q = max(q, p[i] + cut_log(p, n - i))
    m[n] = q
    return q
#

File: codewars/authoring/dart/memoized_log_cutting.dart
import 'dart:math';

// Your new function as given to you by me, your boss.
Function cutLog = (List p, int n) {
  if (n == 0) return 0;
// Some array to store calculated values
  var resultTable = new List(n + 1);
  for (int i = 0; i <= n; i++) resultTable[i] = 0;
// try all lengths up to 'n'
  for (int j = 1; j <= n; j++) {
//try to cut out one piece along the price table
//and calculate the price as sum of this piece plus the price of 'the rest'
    for (int k = 1; k <= j; k++) {
      // Two nested loops = Θ(n^2)
// no magic: since we start at length 0,
// we will already have calculated the price of 'the rest' in our resultTable
      resultTable[j] = max(resultTable[j], p[k] + resultTable[j - k]);
    }
  }
  return resultTable[n]; // Good luck intern!
};

